File: backend/app/schemas/forecasting.py
from __future__ import annotations

from typing import Any

from pydantic import BaseModel, Field


class ForecastMetricsResponse(BaseModel):
    date_column: str | None
    metrics: list[str]


class ForecastRunRequest(BaseModel):
    targets: list[str] | None = None
    horizon_days: int = 30
    standard_horizons: list[int] | None = None
    model_override: str = "Auto"
    granularity: str = "native"
    business_context: str = ""
    # LLM model id for the narrative-interpretation step only — distinct from
    # model_override, which selects the forecasting *engine* (Prophet/ETS/...).
    llm_model: str | None = None


class ForecastOutputOut(BaseModel):
    """Mirrors agents.forecasting.schema.ForecastOutput, but every field the
    error path (a per-target exception in forecast_node) doesn't populate
    gets a default here — a failed target still has to round-trip through
    this model instead of 500ing the whole run."""

    metric: str
    current_value: float = 0.0
    forecasted_value: float = 0.0
    change_percent: float = 0.0
    confidence_score: float | None = None
    forecast_horizon: str = ""
    granularity: str = "daily"
    training_periods: int = 0
    selected_model: str = ""
    model_selection_reason: str = ""
    evaluation: dict[str, Any] = Field(default_factory=dict)
    business_impact: str = "medium"
    business_summary: str = ""
    trend: str = "stable"
    risks: list[str] = Field(default_factory=list)
    opportunities: list[str] = Field(default_factory=list)
    recommended_actions: list[str] = Field(default_factory=list)
    horizons: dict[str, float] = Field(default_factory=dict)
    cv_results: list[dict[str, Any]] = Field(default_factory=list)
    skill_score: float | None = None
    error: str | None = None

    # ── Headline aggregate ──────────────────────────────────────────────
    # "sum" for additive metrics, "mean" for rates — decides whether the
    # horizon figure reads as a total or an average.
    aggregation: str = "sum"
    horizon_total: float | None = None
    horizon_total_lower: float | None = None
    horizon_total_upper: float | None = None
    baseline_window_value: float | None = None
    transform: str = "none"

    # ── Measured uncertainty ────────────────────────────────────────────
    interval_level: float = 0.8
    interval_method: str = ""
    measured_coverage: float | None = None

    # ── What the data supports ──────────────────────────────────────────
    predictability: float | None = None
    trend_strength: float | None = None
    seasonal_strength: float | None = None
    signal_verdict: str = "unknown"

    # ── Data quality / change detection ─────────────────────────────────
    anomalies_detected: int = 0
    anomaly_periods: list[dict[str, Any]] = Field(default_factory=list)
    level_shift: dict[str, Any] | None = None
    data_quality: dict[str, Any] = Field(default_factory=dict)
    series_fingerprint: str = ""
    notes: list[str] = Field(default_factory=list)

    # ── Is it fit to act on? ────────────────────────────────────────────
    reliability: str = "unknown"
    reliability_headline: str = ""
    reliability_reasons: list[str] = Field(default_factory=list)
    recommended_granularity: str = ""
    granularity_reason: str = ""

    # How the metric has held up against forecasts issued on earlier runs.
    drift: dict[str, Any] | None = None


class ForecastRunResponse(BaseModel):
    run_id: str
    forecastable: bool
    validation_message: str
    date_column: str
    frequency: str
    history_length: int
    outputs: list[ForecastOutputOut]
    figures: dict[str, dict[str, Any]]
    report_md: str

File: frontend/src/lib/queries/forecasting.ts
'use client'

import { useMutation, useQuery } from '@tanstack/react-query'
import { useApi } from '@/lib/api/use-api'

export type ForecastMetrics = {
  date_column: string | null
  metrics: string[]
}

export function useForecastMetrics(projectId: string) {
  const { call, ready } = useApi()
  return useQuery({
    queryKey: ['forecast-metrics', projectId],
    queryFn: () => call<ForecastMetrics>(`/api/v1/projects/${projectId}/forecast/metrics`),
    enabled: ready && Boolean(projectId),
  })
}

export type ForecastOutput = {
  metric: string
  current_value: number
  forecasted_value: number
  change_percent: number
  confidence_score: number | null
  forecast_horizon: string
  granularity: string
  training_periods: number
  selected_model: string
  model_selection_reason: string
  evaluation: { mae?: number; rmse?: number; mape?: number; mase?: number }
  business_impact: string
  business_summary: string
  trend: string
  risks: string[]
  opportunities: string[]
  recommended_actions: string[]
  horizons: Record<string, number>
  cv_results: Record<string, any>[]
  skill_score: number | null
  error: string | null

  /** "sum" for additive metrics (a horizon figure is a total), "mean" for rates. */
  aggregation: string
  /** The headline business number: the total/average over the horizon, with its band. */
  horizon_total: number | null
  horizon_total_lower: number | null
  horizon_total_upper: number | null
  /** Same aggregate over the equivalent trailing window — what change_percent compares against. */
  baseline_window_value: number | null
  transform: string

  interval_level: number
  interval_method: string
  /** Leave-one-out coverage of the band on the back-test. Far from interval_level = don't trust the range. */
  measured_coverage: number | null

  /** Share of variation that is structure rather than period-to-period noise. */
  predictability: number | null
  trend_strength: number | null
  seasonal_strength: number | null
  signal_verdict: string

  anomalies_detected: number
  anomaly_periods: { index: number; z: number; date?: string }[]
  level_shift: { index: number; score: number } | null
  data_quality: Record<string, any>
  series_fingerprint: string
  notes: string[]

  /** Whether this forecast is fit to act on. Drives the banner in the UI. */
  reliability: 'reliable' | 'indicative' | 'unreliable' | 'unknown'
  reliability_headline: string
  reliability_reasons: string[]
  recommended_granularity: string
  granularity_reason: string

  /** How the previous run's forecast held up against what actually happened. */
  drift: {
    status: string
    message: string
    compared_points: number
    live_mae: number | null
    expected_mae: number | null
    error_ratio: number | null
    coverage: number | null
    bias: number | null
  } | null
}

export type ForecastRunResult = {
  run_id: string
  forecastable: boolean
  validation_message: string
  date_column: string
  frequency: string
  history_length: number
  outputs: ForecastOutput[]
  figures: Record<string, { data: any[]; layout: Record<string, any> }>
  report_md: string
}

export function useRunForecast(projectId: string) {
  const { call } = useApi()
  return useMutation({
    mutationFn: (input: {
      targets?: string[]
      horizon_days: number
      granularity: string
      model_override: string
      business_context?: string
    }) => call<ForecastRunResult>(`/api/v1/projects/${projectId}/forecast/run`, { method: 'POST', json: input }),
  })
}
